src/puppetlabs/puppetdb/command/dlo.clj

```diff
@@ -1,210 +1,173 @@
-                                     :let [files (compressible-files subdir threshold)]
-                                     :when (seq files)]
-                                 [subdir files]))]
-    (time! (global-metric :compression)
-           (doseq [[subdir files] subdir-files]
-             (compress-subdir! subdir files)))))
+  (:require
+   [clojure.java.io :as io]
+   [clojure.string :as str]
+   [puppetlabs.i18n.core :refer [trs]]
+   [puppetlabs.kitchensink.core :refer [timestamp]]
+   [puppetlabs.puppetdb.nio :refer [copts copt-atomic copt-replace]]
+   [puppetlabs.puppetdb.queue :refer [cmdref->entry parse-metadata]]
+   [metrics.gauges :refer [gauge-fn]]
+   [stockpile :as stock])
+  (:import
+   [java.nio.file AtomicMoveNotSupportedException
+    FileAlreadyExistsException Files LinkOption]
+   [java.nio.file.attribute FileAttribute]))
+
+(defn- parse-cmd-filename [s]
+  (let [rx #"([0-9+])-(.*)"]
+    (when-let [[_ id qmeta] (re-matches rx s)]
+      (parse-metadata qmeta))))
+
+(defn- update-cmd-stats
+  "Updates stats to reflect the receipt of the named command."
+  [stats command size]
+  (let [addnp (fn [v n] (+ (or v 0) n))]
+    (-> stats
+        (update-in ["all" :count] inc)
+        (update-in ["all" :size] addnp size)
+        (update-in [command :count] addnp 1)
+        (update-in [command :size] addnp size))))
+
+(defn- gather-stats [dlo]
+  "Gathers statistics about the filenames directly inside the dlo
+  directory (a Path).  Does not recurse.  e.g.:
+    {\"all\" {:count 104 :size 2073951}
+     \"replace facts\" {:count 34 :size 302116}
+     \"replace catalog\" {:count 19 :size 708182}
+     ...}"
+  (with-open [path-stream (Files/newDirectoryStream dlo)]
+    (reduce (fn [stats p]
+              ;; Assume the trailing .json here and in
+              ;; entry-cmd-err-filename below.
+              (if-let [cmeta (and (.endsWith p ".json")
+                                  (parse-cmd-filename (-> p .getFileName str)))]
+                (update-cmd-stats (:command cmeta) (Files/size p))
+                stats))
+            {"all" {:count 0 :size 0}}
+            path-stream)))
+
+(defn- cmd-gauges [dlo category]
+  "Adds gauges to the dlo for the given category (e.g. \"all\"
+  \"replace command\").  The gauges will pull their values from the
+  dlo atom's :stats."
+  (let [registry (:registry @dlo)
+        stat (fn [what] (get-in (:stats @dlo) what))]
+    {:count (gauge-fn registry ["puppetlabs.puppetdb.dlo" category "count"]
+                      #(stat [category :count]))
+     :size (gauge-fn registry ["puppetlabs.puppetdb.dlo" category "size"]
+                     #(stat [category :size]))}))
+
+(defn- note-new-discard! [dlo command size]
+  ;; Swap twice so that :stats will already have the values that the
+  ;; gauges are going to refer to.
+  (swap! dlo update :stats update-cmd-stats command size)
```

Comment:
Not sure I follow yet -- since the gauges refer to the stat from the "top", I wanted the stats to be fully in place before we create the gauges that refer to them, and I wanted the gauges to refer to the stats by "get-in" lookup, not by pointer reference.